File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_1_A_InsertionSort/ALDS1_1_A.c
#include <stdio.h>

void printArray(int* as, int n)
{
  int i;
  for (i = 0; i < (n - 1); i++) {
    printf("%d ", as[i]);
  }
  printf("%d\n", as[n - 1]);
}

void insertionSort(int* as, int n)
{
  int i, j, k;
  int key;

  printArray(as, n);

  for (i = 1; i < n; i++) {
    key = as[i];
    j = i - 1;
    while (j >= 0 && as[j] > key) {
      as[j + 1] = as[j];
      j--;
    }
    as[j + 1] = key;
    printArray(as, n);
  }
}

int main(int argc, char *argv[])
{
  char row[4];
  int n = atoi(fgets(row, sizeof(row), stdin));
  int as[n];
  int i;

  for (i = 0; i < n; i++) {
    scanf("%d", &as[i]);
  }

  insertionSort(as, n);

  return 0;
}

File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_1_A_InsertionSort/ALDS1_1_A.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func insertionSort(n int, as []int) {
	printArray(as)
	for i := 1; i < n; i++ {
		v := as[i]
		j := i - 1
		for ; j >= 0; j-- {
			if v > as[j] {
				break
			}
			as[j+1] = as[j]
		}
		as[j+1] = v
		printArray(as)
	}
}

func printArray(as []int) {
	fmt.Println(strings.Trim(fmt.Sprint(as), "[]"))
}

func nextInt(sc *bufio.Scanner) int {
	sc.Scan()
	n, err := strconv.Atoi(sc.Text())
	if err != nil {
		panic(err)
	}
	return n
}

func main() {
	sc := bufio.NewScanner(os.Stdin)
	sc.Split(bufio.ScanWords)
	n := nextInt(sc)
	as := make([]int, n)
	for i := 0; i < n; i++ {
		as[i] = nextInt(sc)
	}
	insertionSort(n, as)
}
